src/web/middleware.rs

```diff
@@ -0,0 +1,145 @@
+//! # Web Middleware
+//!
+//! Matches the [Sync Storage middleware](https://github.com/mozilla-services/server-syncstorage/blob/master/syncstorage/tweens.py) (tweens).
+use actix_web::middleware::{Middleware, Response, Started};
+use actix_web::{error::ErrorInternalServerError, http::header, HttpRequest, HttpResponse, Result};
+use chrono::Utc;
+
+/// Default Timestamp used for WeaveTimestamp middleware.
+struct DefaultWeaveTimestamp(f64);
+
+/// Middleware to set the X-Weave-Timestamp header on all responses.
+pub struct WeaveTimestamp;
+
+impl<S> Middleware<S> for WeaveTimestamp {
+    /// Set the `DefaultWeaveTimestamp` and attach to the `HttpRequest`
+    fn start(&self, req: &HttpRequest<S>) -> Result<Started> {
+        // Get millisecond resolution and convert to seconds
+        let ts = Utc::now().timestamp_millis() as f64 / 1_000.0;
+        req.extensions_mut().insert(DefaultWeaveTimestamp(ts));
+        Ok(Started::Done)
+    }
+
+    /// Method is called when handler returns response,
+    /// but before sending http message to peer.
+    fn response(&self, req: &HttpRequest<S>, mut resp: HttpResponse) -> Result<Response> {
+        let extensions = req.extensions();
+        let ts = match extensions.get::<DefaultWeaveTimestamp>() {
+            Some(ts) => ts,
+            None => return Ok(Response::Done(resp)),
+        };
+
+        let weave_ts = if let Some(val) = resp.headers().get("X-Last-Modified") {
+            let resp_ts = val
+                .to_str()
+                .map_err(|_| ErrorInternalServerError("Unable to extract header value"))?
+                .parse::<f64>()
+                .map_err(|_| ErrorInternalServerError("Invalid last-modified set"))?;
+            if resp_ts > ts.0 {
```

Comment:
nit: could use cmp::max

Reply:
Actually, cannot, floats only implement PartialOrd, I tried that first. :grinning: